src/Panel/Ui/Button/VersionsButton.php

```diff
@@ -21,37 +21,94 @@ class VersionsButton extends ViewButton
 {
 	public function __construct(
 		ModelWithContent $model,
-		VersionId|string $versionId = 'latest'
+		public string $mode = 'latest'
 	) {
-		$versionId = $versionId === 'compare' ? 'compare' : VersionId::from($versionId)->value();
-		$viewUrl   = $model->panel()->url(true) . '/preview';
-
 		parent::__construct(
+			model: $model,
 			class: 'k-versions-view-button',
-			icon: $versionId === 'compare' ? 'layout-columns' : 'git-branch',
-			options: [
-				[
-					'label'   => I18n::translate('version.latest'),
-					'icon'    => 'git-branch',
-					'link'    => $viewUrl . '/latest',
-					'current' => $versionId === 'latest'
-				],
-				[
-					'label'   => I18n::translate('version.changes'),
-					'icon'    => 'git-branch',
-					'link'    => $viewUrl . '/changes',
-					'current' => $versionId === 'changes'
-				],
-				'-',
-				[
-					'label'   => I18n::translate('version.compare'),
-					'icon'    => 'layout-columns',
-					'link'    => $viewUrl . '/compare',
-					'current' => $versionId === 'compare'
-				],
+			icon: $this->icon(),
+			text: $this->i18n('version.' . $this->mode()),
+		);
+	}
+
+	/**
+	 * Returns the button icon based on the view's mode
+	 * @since 5.1.0
+	 */
+	public function icon(): string
```

Comment:
`Question`: Should this methods that is called in the constructor actually be `public`? Public methods overshadow `Component::__call()`, so `$versionsButton->icon()` for example will always return the dynamic value based on the mode and not the prop value. As the `$icon` property is `public` as well, one could set the property to a different value on the instance, creating a discrepancy between the property and return value of the method.

In this case, this likely won't cause any problems, I'm asking more about the general coding pattern for such component classes.

Reply:
Yes, that is definitely true. I have used this pattern as well and generally it has worked quite well.

I think in this case however, we don't need to make the method public to test it. If we make it `protected`, we can still call `$versionsButton->icon()`, but PHP will then automatically fall back to the magic `__call()` method, thus accessing the `$icon` property that we want to test.